Unstable test

updateShouldReturnUpdatedWhenRemovePartialValidate(GuiceJamesServer) com.linagora.tmail.james.MemoryPublicAssetSetSetMethodTest
Failing for the past 1 build since #1
Took 1.3 sec
Error Details￼
JsResultException(errors:List((,List(JsonValidationError(List('blobId' is undefined on object. Available keys are 'type', 'status', 'detail'),List())))))
Stack Trace￼
play.api.libs.json.JsResultException: JsResultException(errors:List((,List(JsonValidationError(List('blobId' is undefined on object. Available keys are 'type', 'status', 'detail'),List())))))
	at play.api.libs.json.JsReadable.$anonfun$as$2(JsReadable.scala:25)
	at play.api.libs.json.JsError.fold(JsResult.scala:66)
	at play.api.libs.json.JsReadable.as(JsReadable.scala:24)
	at play.api.libs.json.JsReadable.as$(JsReadable.scala:23)
	at play.api.libs.json.JsUndefined.as(JsLookup.scala:197)
	at com.linagora.tmail.james.common.PublicAssetSetMethodContract.uploadAsset(PublicAssetSetMethodContract.scala:91)
	at com.linagora.tmail.james.common.PublicAssetSetMethodContract.createPublicAssetIdWithIdentityId(PublicAssetSetMethodContract.scala:2001)
	at com.linagora.tmail.james.common.PublicAssetSetMethodContract.updateShouldReturnUpdatedWhenRemovePartialValidate(PublicAssetSetMethodContract.scala:1365)
	at java.base/java.lang.reflect.Method.invoke(Method.java:565)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1604)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1604)
Standard Output￼
10:56:32.647 [ERROR] o.a.j.j.r.UploadRoutes - Unexpected error upon upload /upload/29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6
java.lang.InterruptedException: null
	at java.base/java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:1767)
	at reactor.core.publisher.BlockingIterable$SubscriberIterator.hasNext(BlockingIterable.java:186)
	at java.base/java.util.Spliterators$IteratorSpliterator.tryAdvance(Spliterators.java:1949)
	at java.base/java.util.Spliterators$1Adapter.hasNext(Spliterators.java:669)
	at org.apache.james.util.ReactorUtils$StreamInputStream.nextNonEmptyBuffer(ReactorUtils.java:175)
	at org.apache.james.util.ReactorUtils$StreamInputStream.read(ReactorUtils.java:156)
	at java.base/java.io.FilterInputStream.read(FilterInputStream.java:119)
	at org.apache.commons.fileupload.util.LimitedInputStream.read(LimitedInputStream.java:159)
	at com.google.common.io.CountingInputStream.read(CountingInputStream.java:63)
	at java.base/java.io.InputStream.readNBytes(InputStream.java:411)
	at java.base/java.io.InputStream.readAllBytes(InputStream.java:348)
	at org.apache.james.blob.api.BlobStoreDAO$InputStreamBlob.asBytes(BlobStoreDAO.java:193)
	at org.apache.james.blob.memory.MemoryBlobStoreDAO.lambda$save$0(MemoryBlobStoreDAO.java:75)
	at reactor.core.publisher.MonoCallable.call(MonoCallable.java:72)
	at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap(FluxFlatMap.java:128)
	at reactor.core.publisher.MonoFlatMap.subscribeOrReturn(MonoFlatMap.java:53)
	at reactor.core.publisher.Mono.subscribe(Mono.java:4560)
	at reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.subscribeNext(MonoIgnoreThen.java:265)
	at reactor.core.publisher.MonoIgnoreThen.subscribe(MonoIgnoreThen.java:51)
	at reactor.core.publisher.Mono.subscribe(Mono.java:4576)
	at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap(FluxFlatMap.java:205)
	at reactor.core.publisher.MonoFlatMap.subscribeOrReturn(MonoFlatMap.java:53)
	at reactor.core.publisher.Mono.subscribe(Mono.java:4560)
	at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap(FluxFlatMap.java:202)
	at reactor.core.publisher.MonoFlatMap.subscribeOrReturn(MonoFlatMap.java:53)
	at reactor.core.publisher.Mono.subscribe(Mono.java:4560)
	at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap(FluxFlatMap.java:202)
	at reactor.core.publisher.MonoFlatMap.subscribeOrReturn(MonoFlatMap.java:53)
	at reactor.core.publisher.Mono.subscribe(Mono.java:4560)
	at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap(FluxFlatMap.java:202)
	at reactor.core.publisher.MonoFlatMap.subscribeOrReturn(MonoFlatMap.java:53)
	at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:63)
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:165)
	at reactor.core.publisher.MonoUsing$MonoUsingSubscriber.onNext(MonoUsing.java:231)
	at reactor.core.publisher.FluxSwitchIfEmpty$SwitchIfEmptySubscriber.onNext(FluxSwitchIfEmpty.java:74)
	at reactor.core.publisher.MonoNext$NextSubscriber.onNext(MonoNext.java:82)
	at reactor.core.publisher.FluxConcatMapNoPrefetch$FluxConcatMapNoPrefetchSubscriber.innerNext(FluxConcatMapNoPrefetch.java:259)
	at reactor.core.publisher.FluxConcatMap$ConcatMapInner.onNext(FluxConcatMap.java:865)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:129)
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.secondComplete(MonoFlatMap.java:245)
	at reactor.core.publisher.MonoFlatMap$FlatMapInner.onNext(MonoFlatMap.java:305)
	at reactor.core.publisher.FluxSubscribeOnCallable$CallableSubscribeOnSubscription.run(FluxSubscribeOnCallable.java:252)
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:68)
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:28)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:328)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:309)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:614)
	at java.base/java.lang.Thread.run(Thread.java:1474)

--- a/tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/PublicAssetSetMethodContract.scala
+++ b/tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/PublicAssetSetMethodContract.scala
@@ -49,6 +49,7 @@ import org.junit.jupiter.api.{BeforeEach, Tag, Test}
 import play.api.libs.json.{JsString, Json}
 import reactor.core.scala.publisher.SMono
 
+import scala.annotation.tailrec
 import scala.jdk.CollectionConverters._
 
 object PublicAssetSetMethodContract {
@@ -77,20 +78,33 @@ trait PublicAssetSetMethodContract {
 
   private def uploadAsset(content: Array[Byte] = "Your asset content here".getBytes,
                           contentType: String = "image/png"): UploadResponse = {
-    val uploadResponse: String = `given`()
-      .basePath("")
-      .contentType(contentType)
-      .body(content)
-    .when
-      .post(s"/upload/$ACCOUNT_ID")
-    .`then`
-      .extract
-      .body
-      .asString
+    // The upload occasionally fails server side with a transient InterruptedException while the
+    // blob store reads the request body, returning a problem object instead of the blobId.
+    // Retry a few times to keep this contract test stable.
+    @tailrec
+    def attempt(remaining: Int): UploadResponse = {
+      val uploadResponse: String = `given`()
+        .basePath("")
+        .contentType(contentType)
+        .body(content)
+      .when
+        .post(s"/upload/$ACCOUNT_ID")
+      .`then`
+        .extract
+        .body
+        .asString
 
-    UploadResponse(blobId = (Json.parse(uploadResponse) \ "blobId").as[String],
-      contentType = org.apache.james.mailbox.model.ContentType.of((Json.parse(uploadResponse) \ "type").as[String]).mimeType(),
-      size = (Json.parse(uploadResponse) \ "size").as[Long])
+      (Json.parse(uploadResponse) \ "blobId").asOpt[String] match {
+        case Some(blobId) =>
+          UploadResponse(blobId = blobId,
+            contentType = org.apache.james.mailbox.model.ContentType.of((Json.parse(uploadResponse) \ "type").as[String]).mimeType(),
+            size = (Json.parse(uploadResponse) \ "size").as[Long])
+        case None if remaining > 0 => attempt(remaining - 1)
+        case None => throw new IllegalStateException(s"Upload did not return a blobId: $uploadResponse")
+      }
+    }
+
+    attempt(remaining = 3)
   }
 
   @Test

